### DISCUSSION TITLE
Caching ignores `schema_translate_map`

### DISCUSSION THREAD

**Discussion by jellevwezel** [sqlalchemy/sqlalchemy#10024] [Usage Questions] [answered]
Hi, I'm writing code for multi tenancy (multi schema with one dataase) and I found the following behavior.
It seems that the `compiled_cache` ignores the `schema_translate_map` in some cases.

Version: 2.0.16
Python: 3.10.10
Database: MariaDB 10.x.x
Driver: Asyncmy 0.2.8

My code:

```python

# SharedModel => MetaData(schema='shared')
# TenantModel => MetaData(schema='tenant')


engine = create_async_engine('<db uri>', pool_pre_ping=True)
session_factory = async_sessionmaker(autocommit=False, autoflush=False, expire_on_commit=False)

schema_translate_map = {'shared': 'shared'}
bind =  engine.execution_options(schema_translate_map=schema_translate_map)

async with session_factory(bind=bind) as session:
    try:
        # first database call
        await session.get(TenantModel, 1)  # Should not have access to the tenant here and raises as expected
    except:
        pass

# Then configure a new bind
schema_translate_map = {'shared': 'shared', 'tenant': 'tenant_1'}
bind =  engine.execution_options(schema_translate_map=schema_translate_map)

async with session_factory(bind=bind) as session:
    # second database call
    await session.get(TenantModel, 1)  
    # Should have access to the tenant here but raises error: trying to select from schema 'tenant'
    # Ignoring the new schema_translate_map: tenant => tenant_1
```

Actual error I got:
```
E   sqlalchemy.exc.ProgrammingError: (asyncmy.errors.ProgrammingError) (1146, "Table 'tenant.test_table' doesn't exist")
E   [SQL: SELECT tenant.test_table.id, tenant.test_table.test_col 
E   FROM tenant.test_table]
E   (Background on this error at: https://sqlalche.me/e/20/f405)
```

The error only occurs when first calling the same query with a faulty `schema_translate_map`.
When I disable the caching: i.e. `compiled_cache=None` it works without raising in the second call.
Also, when I supply a cache per tenant, it also works without raising in the second call.

Furthermore, it only happens when the first call raises. I.e. this works fine without touching the cache:

```python

# call to tenant 1
schema_translate_map = {'shared': 'shared', 'tenant': 'tenant_1'}
bind =  engine.execution_options(schema_translate_map=schema_translate_map)

async with session_factory(bind=bind) as session:
    model = await session.get(TenantModel, 1)
    assert model.some_test_col == 'from tenant 1'

# call to tenant 2
schema_translate_map = {'shared': 'shared', 'tenant': 'tenant_2'}
bind =  engine.execution_options(schema_translate_map=schema_translate_map)

async with session_factory(bind=bind) as session:
    await session.get(TenantModel, 1)
    assert model.some_test_col == 'from tenant 2'
```

So it seems to me the caching does not take the schema_translate_map into account when an error occurred, causing this issue.
I don't know if this is a bug or desired behavior or if I'm doing something wrong, that's why I put it in as a discussion post.
Please let me know :)

**Comment by zzzeek** [comment, marked as answer]
hi - 

the problem is that in changing the schema translate map you are adding a key that was not there previously, and the compiler has already compiled the statement with the assumption that the only name which should be "escaped" is the name "shared".  so when it compiles a statement that includes "tenant", it leaves the name "tenant" unescaped and therefore it does not respond to a later schema translate map which *does* include this key.

Solution is to use the same keys every time in the schema translate map:

```py
# will work
schema_translate_map = {'shared': 'shared', "tenant": "fake_schema"}

# will not work, because key is missing
# schema_translate_map = {'shared': 'shared'}

```

SQLAlchemy could resolve this automatically in the following ways:

1. the full list of keys in schema_translate_map could be part of the cache key.  We dont do this however because we don't assume that the given schema_translate_map is not a dynamic mapping, and we don't assume this isnt a multitenant system where the map has 10000 keys
2. it could compare the keys each time and somehow invalidate the cache entry, but this would make the cache not function

What SQLAlchemy could do is this:

1. when a compiled object that includes schema translate map is pulled from the cache, we look in the compiled object itself to attain the set of all schema names that were actually "translated" and those that were not "translated"
2. we ensure the map has all keys in the "translated" set and does not have keys that are in the not "translated" set, otherwise it raises

This is a bit burdensome on every request though, and the end result is that this is only a "linter" that will only tell the user to modify their program and try again.

that is, the map here needs to have the same keys, and it looks like your original case is that your statement raised an error because the map was insufficient.   I would assume this is some kind of test suite and delivering insufficient schema translate maps that raise errors are not a regular use case ?

I will add a warning to the documentation now.      






for reference a full demo is below


```py
from __future__ import annotations

import asyncio
from sqlalchemy.ext.asyncio import create_async_engine, async_sessionmaker


from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class TenantModel(Base):
    __tablename__ = "tenant_model"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]

    __table_args__ = {"schema": "tenant"}



engine = create_async_engine('postgresql+asyncpg://scott:tiger@localhost/test', pool_pre_ping=True, echo=True)
session_factory = async_sessionmaker(autocommit=False, autoflush=False, expire_on_commit=False)

REAL_SCHEMA = "test_schema"

# will work
schema_translate_map = {'shared': 'shared', "tenant": "fake_schema"}

# will not work, because key is missing
# schema_translate_map = {'shared': 'shared'}

global_bind =  engine.execution_options(schema_translate_map=schema_translate_map)


async def main():
    async with engine.begin() as conn:
        # create table and data in the desired schema using direct SQL
        # to eliminate any possibilty of incorrect SQL
        await conn.exec_driver_sql(f"DROP TABLE IF EXISTS {REAL_SCHEMA}.tenant_model")
        await conn.exec_driver_sql(f"CREATE TABLE {REAL_SCHEMA}.tenant_model (id SERIAL PRIMARY KEY, data VARCHAR)")
        await conn.exec_driver_sql(f"INSERT INTO {REAL_SCHEMA}.tenant_model (id, data) VALUES (1, 'row one')")
        await conn.commit()

    for i in range(5):
        async with session_factory(bind=global_bind) as session:
            try:
                # first database call
                await session.get(TenantModel, 1)  # Should not have access to the tenant here and raises as expected
            except:
                pass
            else:
                assert False, "should have raised"

        # Then configure a new bind
        sub_schema_translate_map = {'shared': 'shared', 'tenant': REAL_SCHEMA}
        sub_bind =  engine.execution_options(schema_translate_map=sub_schema_translate_map)

        async with session_factory(bind=sub_bind) as session:
            # second database call
            await session.get(TenantModel, 1)


asyncio.run(main())
```

**Reply by zzzeek** [reply]
another way this could go is that we tokenize **all** schema names and provide a fallback within the token.    Maybe I can look into that though it might be a little bit backwards incompatible, if we have an existing error raise for schema token not found.

**Reply by zzzeek** [reply]
> another way this could go is that we tokenize **all** schema names and provide a fallback within the token. Maybe I can look into that though it might be a little bit backwards incompatible, if we have an existing error raise for schema token not found.

^^^^^^^^^^^^^^^^^^^^^^^ :+1: 


there is an opposite issue which is when the map *does* include the key and later does not.  we right now don't even catch for this, which leaves us open here to have it fallback to the missing name.

This is enough to qualify as a bug and is in #10025 will be released in 2.0.18.

**Reply by jellevwezel** [reply]
Hi,
Tested it by just changing the values of the `schema_translate_map`, i.e.:

```python
map_1 = {'tenant': 'NOT_SET', 'shared': 'shared'}
map_2 = {'tenant': 'tenant_1', 'shared': 'shared'}
```
And that worked without a problem.
Also, to answer your question, this came from a test set to check if the tenants were scoped correctly, and was not normal usage.

Thanks for the reply, and I'm glad I could contribute :smile:
